File: internal/testhttp/sse.go
// Package testhttp 为协议测试提供标准 SSE 帧，生产代码不得导入。
package testhttp

import (
	"encoding/json"
	"fmt"
	"net/http"
)

// WriteSSE 把测试结果描述发布成流式协议事件；错误信封仍使用 HTTP JSON 错误格式。
func WriteSSE(w http.ResponseWriter, r *http.Request, payload any) error {
	raw, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	var body map[string]any
	if err = json.Unmarshal(raw, &body); err != nil {
		return err
	}
	if _, ok := body["error"]; ok {
		return json.NewEncoder(w).Encode(body)
	}
	w.Header().Set("Content-Type", "text/event-stream")
	emit := func(value any) error {
		raw, err := json.Marshal(value)
		if err != nil {
			return err
		}
		_, err = fmt.Fprintf(w, "data: %s\n\n", raw)
		if f, ok := w.(http.Flusher); ok {
			f.Flush()
		}
		return err
	}
	if r.URL.Path == "/responses" || r.URL.Path == "/v1/responses" {
		items, _ := body["output"].([]any)
		for index, value := range items {
			item, _ := value.(map[string]any)
			switch item["type"] {
			case "function_call":
				if args, ok := item["arguments"].(string); ok {
					if err := emit(map[string]any{"type": "response.function_call_arguments.delta", "output_index": index, "item_id": item["id"], "delta": args}); err != nil {
						return err
					}
				}
			case "message":
				if parts, ok := item["content"].([]any); ok {
					for _, part := range parts {
						p, _ := part.(map[string]any)
						if text, ok := p["text"].(string); ok {
							if err := emit(map[string]any{"type": "response.output_text.delta", "output_index": index, "item_id": item["id"], "delta": text}); err != nil {
								return err
							}
						}
					}
				}
			}
			if err := emit(map[string]any{"type": "response.output_item.done", "output_index": index, "item": item}); err != nil {
				return err
			}
		}
		event := "response.completed"
		if body["status"] == "incomplete" {
			event = "response.incomplete"
		}
		if body["status"] == "failed" {
			event = "response.failed"
		}
		return emit(map[string]any{"type": event, "response": body})
	}
	choices, _ := body["choices"].([]any)
	for _, value := range choices {
		choice, _ := value.(map[string]any)
		message, _ := choice["message"].(map[string]any)
		if calls, ok := message["tool_calls"].([]any); ok {
			for i, c := range calls {
				c.(map[string]any)["index"] = i
			}
		}
		if err := emit(map[string]any{"id": body["id"], "object": "chat.completion.chunk", "model": body["model"], "choices": []any{map[string]any{"index": choice["index"], "delta": message, "finish_reason": choice["finish_reason"]}}, "usage": body["usage"]}); err != nil {
			return err
		}
	}
	_, err = fmt.Fprint(w, "data: [DONE]\n\n")
	return err
}

File: scripts/swe_test_runner/sse.py
"""SWE Test Runner 的有界 SSE 能力探针解码，不提供非流式降级。"""

import json


def events(response):
    """按 SSE 空行分隔事件，保留 UTF-8 和多行 data。"""
    data = []
    size = 0
    total = 0
    for line in response:
        if isinstance(line, bytes):
            line = line.decode("utf-8")
        line = line.rstrip("\r\n")
        size += len(line.encode("utf-8"))
        total += len(line.encode("utf-8"))
        if total > 4 * 1024 * 1024:
            raise RuntimeError("SSE 探针响应超过 4MiB")
        if size > 1024 * 1024:
            raise RuntimeError("SSE 探针事件超过 1MiB")
        if not line:
            if data:
                yield "\n".join(data)
            data, size = [], 0
        elif line.startswith("data:"):
            data.append(line[5:].removeprefix(" "))
    if data:
        raise RuntimeError("SSE 探针在事件结束前断流")


def decode_probe(response, protocol):
    """只在收到协议终止标记后返回可供 nonce 校验的完整结果。"""
    output = {}
    text = ""
    calls = {}
    finish = None
    for raw in events(response):
        if raw == "[DONE]":
            if protocol != "chat_completions" or finish not in {"stop", "tool_calls", "length", "content_filter"}:
                raise RuntimeError("SSE 探针缺少合法完成原因")
            return {"choices": [{"index": 0, "finish_reason": finish,
                                  "message": {"role": "assistant", "content": text,
                                              "tool_calls": [calls[k] for k in sorted(calls)]}}]}
        event = json.loads(raw)
        if not isinstance(event, dict):
            raise RuntimeError("SSE 探针事件不是对象")
        if event.get("error") or event.get("type") in {"error", "response.failed"}:
            raise RuntimeError("SSE 探针返回协议失败事件")
        if protocol == "responses":
            kind = event.get("type")
            if kind == "response.output_item.done":
                index = event.get("output_index")
                if not isinstance(index, int) or index < 0 or index in output:
                    raise RuntimeError("SSE 探针输出项身份无效或重复")
                output[index] = event.get("item")
            elif kind == "response.completed":
                result = event.get("response") or {}
                if result.get("status") != "completed" or not output:
                    raise RuntimeError("SSE 探针缺少完整输出项")
                if sorted(output) != list(range(len(output))):
                    raise RuntimeError("SSE 探针输出项序号不连续")
                return {**result, "output": [output[k] for k in sorted(output)]}
            elif kind == "response.incomplete":
                raise RuntimeError("SSE 探针响应被截断")
        elif protocol == "chat_completions":
            for choice in event.get("choices") or []:
                if choice.get("index", 0) != 0:
                    raise RuntimeError("SSE 探针只允许一个候选回复")
                delta = choice.get("delta") or {}
                text += delta.get("content") or ""
                for part in delta.get("tool_calls") or []:
                    index = part.get("index")
                    if not isinstance(index, int) or index < 0:
                        raise RuntimeError("SSE 探针工具序号无效")
                    call = calls.setdefault(index, {"id": "", "type": "function", "function": {"name": "", "arguments": ""}})
                    call["id"] += part.get("id") or ""
                    function = part.get("function") or {}
                    call["function"]["name"] += function.get("name") or ""
                    call["function"]["arguments"] += function.get("arguments") or ""
                if choice.get("finish_reason") is not None:
                    finish = choice["finish_reason"]
        else:
            raise RuntimeError("未知 SSE 探针协议")
    raise RuntimeError("SSE 探针未收到完整终止事件")
